### DISCUSSION TITLE
get_calendar_events returns an empty object

### DISCUSSION THREAD

**Discussion by bbeck6** [ucfopen/canvasapi#615] [Q&A] [answered]
Hi all,
Maybe someone can help. I can't tell if this is a bug or I'm using the method incorrectly. When I call the following:

`obj_canvas = Canvas(API_URL, API_KEY) `
`pl_events = obj_canvas.get_calendar_events()`

the returned object pl_events (a paginated list of calendar events) contains no data. i.e., pl_events[0] results in an out-of-range error. 

I've tried everything I can think of as an input argument, but this function never returns anything. Can someone give an example of how to make it work? All of my other API code on the obj_canvas object is working fine.
Thanks!

**Comment by bennettscience** [comment]
Hey @bbeck6,
It's not a bug with the library. The way the docs read, you would _think_ that if you don't include any arguments, it just gets the whole calendar, but that's not the case. The `start_date` param defaults to today if you don't pass in a value (which makes sense). The `end_date` param _also_ defaults to today (which doesn't make sense to me). So, if you're calling the method without any arguments, it's only looking for events _today_.

I would either call `obj_canvas.get_calendar_events(all_events=True)` to get everything and then filter in your script _or_ call `obj_canvas.get_calendar_events(end_date=future_date)` to get all events coming up beyond today. You can either set `future_date` to some `datetime` object (make sure it's ISO 8601 formatted) or a YYYY-MM-DD string.

**Reply by bbeck6** [reply]
Thanks for the reply!
I did some more testing, trying the `all_events=True` flag only. What I noticed is that this will ONLY send back events associated purely with the 'user' (me), and not events that are tied to a particular 'course' calendar. These are the ones that need to be edited.

I've tried supplying `context_code='course_######'` but this doesn't return anything. Is this intended to work? I'm at a loss as to how to grab calendar events associated with a course. If you could post even one working example, I would be most grateful. 

Thanks!

**Comment by bennettscience** [comment, marked as answer]
You need to use the `context_codes` argument and then pass it course values like this:

`canvas.get_calendar_events(context_codes=["course_12345"])`.

You can pass up to ten contexts per request, so you can get multiple courses or multiple users all at once. The context code format is `resource_id` for courses, users, or groups.

[API docs](https://canvas.instructure.com/doc/api/calendar_events.html#method.calendar_events_api.index)

**Comment by bennettscience** [comment]
Sorry, make sure you context codes are in quotes - it expects a string.

**Reply by bbeck6** [reply]
Yep, that was it! It's 'context_codes' NOT 'context_code' (grrrr), and you were right, the function expects a LIST, even if you have only one code to send. For the record, here's a full, working line of code for anyone who wants to get ALL the calendar events for a single COURSE:
For given Canvas object `obj_canvas`, run:
`pl_events = obj_canvas.get_calendar_events(context_codes=['course_#####'], all_events=True)`

Thanks again, you saved me untold amounts of time!
